version_test.go

```diff
@@ -393,6 +395,72 @@ func TestVersionSegments64(t *testing.T) {
 	}
 }
 
+func TestJsonMarshal(t *testing.T) {
+	cases := []struct {
+		version string
+		err     bool
+	}{
+		{"1.2.3", false},
+		{"1.2.0-x.Y.0+metadata", false},
+		{"1.2.0-x.Y.0+metadata-width-hypen", false},
+		{"1.2.3-rc1-with-hypen", false},
+		{"1.2.3.4", false},
+		{"1.2.0.4-x.Y.0+metadata", false},
+		{"1.2.0.4-x.Y.0+metadata-width-hypen", false},
+		{"1.2.0-X-1.2.0+metadata~dist", false},
+		{"1.2.3.4-rc1-with-hypen", false},
+		{"1.2.3.4", false},
+	}
+
+	for _, tc := range cases {
+
+		v, err := NewVersion(tc.version)
+		if err != nil {
+			t.Errorf("error for version %q: %s", tc.version, err)
+		}
+		parsed, err2 := json.Marshal(v)
+		if err2 != nil {
+			t.Errorf("error marshaling version %q: %s", tc.version, err2)
+		}
+		result := string(parsed)
+		expected := fmt.Sprintf("%q", tc.version)
+		if result != expected && !tc.err {
+			t.Errorf("Error marshaling unexpected marshaled content: result=%q expected=%q", result, expected)
+		}
+	}
+}
+
+func TestJsonUnmarshal(t *testing.T) {
+	cases := []struct {
+		version string
+		err     bool
+	}{
+		{"1.2.3", false},
+		{"1.2.0-x.Y.0+metadata", false},
+		{"1.2.0-x.Y.0+metadata-width-hypen", false},
+		{"1.2.3-rc1-with-hypen", false},
+		{"1.2.3.4", false},
+		{"1.2.0.4-x.Y.0+metadata", false},
+		{"1.2.0.4-x.Y.0+metadata-width-hypen", false},
+		{"1.2.0-X-1.2.0+metadata~dist", false},
+		{"1.2.3.4-rc1-with-hypen", false},
+		{"1.2.3.4", false},
```

Comment:
Not sure what values I can feed in to force a failure here.

Reply:
Hi @Jukie,
Thanks for submitting this PR.  I don't see much benefit in negative tests here because errors are only returned by code that's not specifically under test by this test function (the standard library json package and the parsing by `NewVersion()`).  The code that really needs to be tested in your implementation is the transfer of data from the temporary struct to the receiver, and I think the tests you already have cover that logic.